compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp

```diff
@@ -131,14 +141,63 @@ struct FoldSuccessiveTensorInsertSliceOps
   }
 };
 
+struct GatherFusionPattern : public OpRewritePattern<linalg::YieldOp> {
+  using OpRewritePattern<linalg::YieldOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::YieldOp yieldOp,
+                                PatternRewriter &rewriter) const override {
+    // Specific case. The linalg generic implementation of "gather"
+    // cannot be fused because it there is no producer-consumer
+    // relationship between the two generics. This is because the indexing
+    // is not affine (index values come from a tensor).
+    if (yieldOp->getNumOperands() != 1) {
+      return failure();
+    }
+    auto extractOp = dyn_cast_or_null<tensor::ExtractOp>(
+        yieldOp->getOperand(0).getDefiningOp());
+    if (!extractOp) {
+      return failure();
+    }
+
+    // match the generic that gens a higher bitwidth tensor
+    auto definingOp = dyn_cast_or_null<linalg::GenericOp>(
+        extractOp.getOperand(0).getDefiningOp());
+    if (!definingOp) {
+      return rewriter.notifyMatchFailure(
+          yieldOp, "expected extract operand to be a generic op");
+    }
+
+    // generic body should contain only a arith.extf and a linalg.yield
+    auto &ops = definingOp->getRegion(0).front().getOperations();
+    if (ops.size() != 2 || !isa<arith::ExtFOp>(ops.front()) ||
+        !isa<linalg::YieldOp>(ops.back())) {
+      return rewriter.notifyMatchFailure(yieldOp,
+                                         "expected generic op to have 2 ops");
+    }
+
+    // move definingOp's body just before the yield op (root)
+    rewriter.inlineBlockBefore(
```

Comment:
This works? What about the yield operation of the defining operation?
 EDIT: Oh I see it below... That works.